feat: add directory pre call filtering

TomChv opened the pull request:
Fixes #9490

Signed-off-by: Tom Chauveau <tom@epitech.eu>

TomChv commented on core/directory.go:
```diff
@@ -37,6 +37,11 @@ type Directory struct {
+	// The list of excluded paths of that directory.
+	// This is useful for the ignore directive to avoid double exclusion between
+	// CLI call and module call.
+	Excluded []string
```
I'm not sure how to use this yet, I keep a list of ignored pattern but I'm not sure it's worth it.

There's no more `resolveDirectoryFromCaller` so I need an hint on that.

jedevc reviewed (commented):
Looks neat, the only thing is that it's inefficient *if* we've already removed that data from this directory, or loaded it with that exclude pattern (e.g. from the CLI, or ). Don't *think* this is a blocker, but it does add extra file manipulation to the module-call path :thinking: cc @helderco I know you mentioned this previously.

As a potential solution for this:
- Create a new property (similar to `Excluded` you've done, but that has included as well) on `Directory` that can be set by `Host.Directory` that indicates the include patterns from when it was loaded (we should convert exclude patterns to include patterns - I *think* we do this somewhere else?).
- When we do `WithFile`/`WithDirectory`/create any new files, then we update the include patterns for the directory with the `dest` path (and append the `include`/`exclude` paths, using `filepath.Join`).
	- We could also do the same for `Without` and add an inverse pattern.
- Then when we do `applyIgnoreOnDir`, we can look at the directory's  patterns, and see if they differ from the `defaultPath` on the arg. If they match, we've already removed them, so we can skip this. Otherwise, we'll need to apply the filter.

I *think* this should work?

TomChv commented on core/modfunc.go:
```diff
@@ -153,6 +153,13 @@ func (fn *ModuleFunction) setCallInputs(ctx context.Context, opts *CallOpts) ([]
 		}
 
+		if len(arg.metadata.Ignore) > 0 {
+			converted, err = fn.applyIgnoreOnDir(ctx, opts.Server, arg.metadata, converted)
```
I'm not sure it's working as expected, I successfully get a `core.Directory` but I end up with the following error:

```
Error: failed to generate code: input: moduleSource.withName.withSDK failed to load sdk for module source: failed to get sdk            
sdk module python-sdk: select: failed to set call inputs: failed to convert arg "sdkSourceDir": *core.NullableType.ConvertTo            
pected Derefable, got dagql.Instance[*github.com/dagger/dagger/core.Directory]:
```

Do you have any idea on how I could fix that?

TomChv commented on core/directory.go:
```diff
@@ -514,6 +520,7 @@ func (dir *Directory) WithDirectory(ctx context.Context, destDir string, src *Di
 		return nil, err
 	}
 
+	dir.Excluded = append(dir.Excluded, filter.Exclude...)
```
Because include is only set through the `withDirectory` API call, I expect people to use it smart and not use negative pattern there.
It makes sense to use negatives for `exclude` only.

However, buildkit seems smart enough to resolve it correctly so we could ideally save it as it's defined by the call.

TomChv commented:
> Looks neat, the only thing is that it's inefficient _if_ we've already removed that data from this directory, or loaded it with that exclude pattern (e.g. from the CLI) - see the perf concern here: [#8436 (comment)](https://github.com/dagger/dagger/pull/8436#discussion_r1768373644). Don't _think_ this is a blocker, but it does add extra file manipulation to the module-call path 🤔 cc @helderco I know you mentioned this previously.
> 
> As a potential solution for this:
> 
> * Create a new property (similar to `Excluded` you've done, but that has included as well) on `Directory` that can be set by `Host.Directory` that indicates the include patterns from when it was loaded (we should convert exclude patterns to include patterns - I _think_ we do this somewhere else?).
> * When we do `WithFile`/`WithDirectory`/create any new files, then we update the include patterns for the directory with the `dest` path (and append the `include`/`exclude` paths, using `filepath.Join`).
>   
>   * We could also do the same for `Without` and add an inverse pattern.
> * Then when we do `applyIgnoreOnDir`, we can look at the directory's  patterns, and see if they differ from the `defaultPath` on the arg. If they match, we've already removed them, so we can skip this. Otherwise, we'll need to apply the filter.
> 
> I _think_ this should work?

I've been thinking about it and tried different things but I'm not sure it would work.

Imagine you have an arg:

```go
//+ignore=["**", "!main.go"]
dir *dagger.Directory
...
```

Whatever happens on the directory, like any addition using `withDirectory` shouldn't change the way the ignore directive is done.
It should always ignore everything example `main.go` so updating the exclude/include pattern isn't gonna work.

What we basically want to avoid with this double exclusion only concerns arguments sent from the CLI.
Like

```
dagger call --dir .
```
The ignore directive is already applied by the CLI so we shouldn't duplicate it, but that's the only case where we actually want to *not* apply the ignore.

So maybe, what I could do instead is just add an optional flag on the `Ignored` on `host.WithDirectory` so we know that the ignore pattern has already been done for this case. And set it to false after the input is set (because a deeper function may have a specific ignore directive that we want to apply after that.
What do you think of that? It's a much simpler solution and it targets the specific problem.

TomChv commented on core/directory.go:
```diff
@@ -356,7 +362,7 @@ func (dir *Directory) Glob(ctx context.Context, pattern string) ([]string, error
 			// HACK: ideally, we'd have something like MatchesExact, which
 			// would skip the parent behavior that we don't really want here -
-			// oh well, let's just fake it with false
+			// oh well, let's just fake it wi- false
```
Fixed by [7e031a7](https://github.com/dagger/dagger/pull/9955/commits/7e031a719fbaa3e61f96d7ce0c96cbe00498b172)
